File: src/com/nerdscorner/mvp/events/utils/GradleUtils.java
package com.nerdscorner.mvp.events.utils;

import com.intellij.openapi.vfs.VirtualFile;

import java.io.BufferedReader;
import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStreamReader;

import javax.annotation.Nullable;

public final class GradleUtils {
    public static final String LIBRARY_DEPENDENCY_PACKAGE = "com.nerdscorner.mvp:";

    public static final String MVP_LIB_INTERFACES_DEPENDENCY = "'com.nerdscorner.mvp:interfaces:1.1.0'";
    public static final String MVP_LIB_EVENTS_DEPENDENCY = "'com.nerdscorner.mvp:events:1.3.0'";

    private static final String COMPILE = "\tcompile ";
    private static final String DEPENDENCIES_BLOCK_BEGIN = "dependencies {";
    private static final String DEPENDENCIES_BLOCK_BEGIN_ALT = "dependencies{";

    public static boolean addDependency(VirtualFile rootFolder, String dependency) {
        try {
            VirtualFile appGradleFile = getAppGradleFile(rootFolder);
            String gradleFilegradleFileContent = getGradleFileContent(rootFolder);
            int dependenciesBlockBegin = gradleFilegradleFileContent.indexOf(DEPENDENCIES_BLOCK_BEGIN);
            if (dependenciesBlockBegin == -1) {
                dependenciesBlockBegin = gradleFilegradleFileContent.indexOf(DEPENDENCIES_BLOCK_BEGIN_ALT);
                if (dependenciesBlockBegin == -1) {
                    throw new IllegalStateException("Malformed build.gradle file, dependencies block not recognized. Expected dependencies { ... }");
                }
            }

            final int SEEKING = 0;
            final int DONE = 2;
            BufferedReader reader = new BufferedReader(new InputStreamReader(new ByteArrayInputStream(appGradleFile.contentsToByteArray())));
            String line;
            StringBuilder gradleFileBuilder = new StringBuilder();
            int state = SEEKING;
            while ((line = reader.readLine()) != null) {
                gradleFileBuilder
                        .append(line)
                        .append(System.lineSeparator());
                if (state == DONE) {
                    continue;
                }
                if (line.contains(DEPENDENCIES_BLOCK_BEGIN) || line.contains(DEPENDENCIES_BLOCK_BEGIN_ALT)) {
                    gradleFileBuilder.append(COMPILE).append(dependency).append(System.lineSeparator());
                    state = DONE;
                }
            }
            FileWriter fileWriter = new FileWriter(new File(getAppGradleFile(rootFolder).getPath()));
            fileWriter.write(gradleFileBuilder.toString());
            fileWriter.close();
            return true;
        } catch (IOException e) {
            e.printStackTrace();
        }
        return false;
    }

    @Nullable
    public static String getGradleFileContent(VirtualFile rootFolder) {
        VirtualFile appGradleFile = getAppGradleFile(rootFolder);
        try {
            return FileReader.getFileContents(new FileInputStream(appGradleFile.getPath()));
        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }

    private static VirtualFile getAppGradleFile(VirtualFile rootFolder) {
        return rootFolder
                .getParent()
                .getParent()
                .getParent()
                .findChild("build.gradle");
    }

    public static boolean hasDependency(VirtualFile rootFolder, String dependency) {
        return getGradleFileContent(rootFolder).contains(dependency);
    }
}

File: src/com/nerdscorner/mvp/utils/gradle/GradleUtils.kt
package com.nerdscorner.mvp.utils.gradle

import com.intellij.ide.util.PropertiesComponent
import java.io.File
import java.io.FileInputStream
import java.io.FileWriter

import com.intellij.openapi.actionSystem.ActionManager
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.vfs.VirtualFile
import com.nerdscorner.mvp.utils.Constants
import com.nerdscorner.mvp.utils.FileReader

object GradleUtils {
    private val propertiesComponent = PropertiesComponent.getInstance()
    private var latestEventsLibVersion = propertiesComponent.getValue(Constants.Properties.PROPERTY_LATEST_EVENTS_LIB_VESION, "1.2.0")
    const val MVP_LIB_EVENTS_DEPENDENCY_PKG = "'com.github.marcherdiego.mvp:events:"

    private var latestCoroutinesLibVersion = propertiesComponent.getValue(Constants.Properties.PROPERTY_LATEST_COROUTINES_LIB_VESION, "1.2.0")
    const val COROUTINES_LIB_EVENTS_DEPENDENCY_PKG = "'com.github.marcherdiego.mvp:coroutines:"

    private const val GRADLE_FILE = "build.gradle"
    private const val ANDROID_SYNC_COMMAND = "Android.SyncProject"

    fun getGradleFileContent(rootFolder: VirtualFile): String? {
        try {
            val appGradleFile = getAppGradleFile(rootFolder)
            return FileReader.getFileContents(FileInputStream(appGradleFile!!.path))
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return null
    }

    fun getAppGradleFile(rootFolder: VirtualFile): VirtualFile? {
        return rootFolder
                .parent
                .parent
                .parent
                .findChild(GRADLE_FILE)
    }

    @JvmStatic
    fun hasDependency(root: VirtualFile, dependency: String) = getGradleFileContent(root)?.contains(dependency) ?: false

    fun restoreGradleFile(savedGradleFile: String?, rootFolder: VirtualFile) {
        if (savedGradleFile == null) {
            return
        }
        try {
            val manifestFile = File(getAppGradleFile(rootFolder)?.path ?: return)
            val fileWriter = FileWriter(manifestFile)
            fileWriter.write(savedGradleFile)
            fileWriter.close()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    @JvmStatic
    fun performSync(actionEvent: AnActionEvent) {
        ActionManager
                .getInstance()
                .getAction(ANDROID_SYNC_COMMAND)
                .actionPerformed(actionEvent)
    }

    fun setLatestEventsLibVersion(version: String?) {
        latestEventsLibVersion = version ?: return
        propertiesComponent.setValue(Constants.Properties.PROPERTY_LATEST_EVENTS_LIB_VESION, latestEventsLibVersion)
    }

    fun setLatestCoroutinesLibVersion(version: String?) {
        latestCoroutinesLibVersion = version ?: return
        propertiesComponent.setValue(Constants.Properties.PROPERTY_LATEST_COROUTINES_LIB_VESION, latestCoroutinesLibVersion)
    }

    fun getLatestEventsLibVersion() = "$MVP_LIB_EVENTS_DEPENDENCY_PKG$latestEventsLibVersion'"

    fun getLatestCoroutinesLibVersion() = "$COROUTINES_LIB_EVENTS_DEPENDENCY_PKG$latestCoroutinesLibVersion'"
}
